File: NVQL-GDB/NVQL-Src/NVQLPathQuery.py
from neo4j import GraphDatabase
from pandas import DataFrame
from datetime import datetime
import time

from sympy import true

class GDBAGGen:
    def __init__(self, uri, user, password):
        try:
            self.driver = GraphDatabase.driver(uri, auth=(user, password))
        except Exception as e:
            print("Failed to initialize Neo4j driver:", e)

    def close(self):
        if self.driver is not None:
            self.driver.close()
    
    def query(self, query, db=None):
        assert self.driver is not None, "Driver is not initialized!"
        session=None
        response=None
        try:
            if db is not None:
                session=self.driver.session(database=db)
            else:
                session=self.driver.session()
            
            response=list(session.run(query))
        except Exception as e:
            print("Query failed:", e)
        finally:
            if session is not None:
                session.close()
        return response

Neo4jDBCon = GDBAGGen("bolt://localhost:7687", "neo4j", "abcdefgh")


HARD_CODED = true

if HARD_CODED:
    srcIP='192.168.148.1'
    dstIP='192.168.148.187'
else:
    print("Enter source IP address: ", end='')
    srcIP = input()
    print("Enter destination IP address: ", end='')
    dstIP = input()


query = "match (src:aghost{ipAddr: '" + srcIP + "'})<-[:AT_HOST]-(gi1:aggi)-[:INSTANCE_OF]->(g:agag{name:'user'}), " + \
        "(dst:aghost{ipAddr: '" + dstIP + "'})<-[:AT_HOST]-(gi2:aggi)-[:INSTANCE_OF]->(g:agag{name:'user'}) " + \
        "match p=shortestPath((gi1)-[:NEXT*1..20]->(gi2)) return nodes(p), length(p)"


print(query)
t1=time.time()
result=Neo4jDBCon.query(query)
t2=time.time()
print("Path query execution time = " + str(t2-t1) + " seconds")
dtf_data = DataFrame([dict(_) for _ in result])
print(dtf_data)

Neo4jDBCon.close()

File: NVQL-GDB/NVQL-Src/NVQLPathQuery.java

import java.io.*;
//import java.net.*;
//import java.nio.*;
import java.sql.*;
//import java.util.*;
//import java.lang.*;
import java.util.Properties;

public class NVQLPathQuery {

    public static void main(String[] args) throws Exception {

        Connection neo4jCon = null;
        long startTime, endTime;
        NVQLQueryExecTime timeobj = new NVQLQueryExecTime();
        String propFileName = "./Resources/config.properties";
        String neo4jPort = "", neo4jPassword = "", neo4jIpAddress = "", neo4jUserName = "";

        NVQLReadPropertiesFileTest readProperties = new NVQLReadPropertiesFileTest();
        try {
            Properties prop = readProperties.readPropertiesFile(propFileName);
            neo4jPort = prop.getProperty("neo4jPort");
            neo4jPassword = prop.getProperty("neo4jPassword");
            neo4jIpAddress = prop.getProperty("neo4jIpAddress");
            neo4jUserName = prop.getProperty("neo4jUserName");
            System.out.println("Reading values from config.properites...");
        } catch (IOException ex) {
            System.out.println(ex.getMessage());
            ex.printStackTrace();
        }

        NVQLInitNeo4jDB t = new NVQLInitNeo4jDB(neo4jIpAddress, neo4jPort, "", neo4jUserName, neo4jPassword);

        neo4jCon = t.getConnection();
        
        startTime=System.nanoTime();

        String query = "match (src:aghost{ipAddr: '192.168.148.1'})<-[:AT_HOST]-(gi1:aggi)-[:INSTANCE_OF]->(g:agag{name:'user'}), " +
        "(dst:aghost{ipAddr: '192.168.148.14'})<-[:AT_HOST]-(gi2:aggi)-[:INSTANCE_OF]->(g:agag{name:'user'})" +
        "match p=shortestPath((gi1)-[:NEXT*1..10]->(gi2)) return nodes(p), length(p";

        ResultSet result = null;
        try {
            PreparedStatement pstmt = neo4jCon.prepareStatement(query);

            result = pstmt.executeQuery();

            System.out.println("NVQL Path Query: Success");

            System.out.println(result);
            
        } catch (Exception e) {
            System.out.println("NVQL Path Query: Failure");
            System.out.println(e);
            System.exit(0);
        }

        endTime=System.nanoTime();
        timeobj.displayTime(startTime,endTime);

    }
}